content/post/2021-07-12-code-detective/index.Rmd

```diff
@@ -0,0 +1,267 @@
+Sometimes print-debugging is the only technique you might be able to use if [non-interactive debugging](https://adv-r.hadley.nz/debugging.html#print-debugging).
+It can also be perfect to know where a loop breaks which motivated the tweet below by Sharla Gelfand:
+
+`r hugodown::embed_tweet("1382090144229044226")`
+
+But often you will have to go experiment under the hood.
+For doing that efficiently you will need to learn about `browser()` and friends!
+Or only just `browser()` to start with!
+
+The basic idea is that you just replace the `print()` command you were about to right with `browser()`, run the code and voilà! You entered the debugger and can run code line by line, explore options and environment variables, etc.
+Over time it'll become a habit of yours, at least that's what happened to me once I saw the light. 😁
+
+Here are some good resources to learn about debugging tools.
+These resources also overlap with some of the objectives of this very blog post.
+
+- [Debugging advice in the Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/debugging.html);
+
+- [Webinar debugging techniques in RStudio by Amanda Gadrow](https://www.rstudio.com/resources/webinars/debugging-techniques-in-rstudio/);
+
+- [Debugging advice in the Hands-on programming with R book by Garrett Grolemund](https://rstudio-education.github.io/hopr/debug.html);
+
+- [Debugging advice in the materials of the course "What they forgot to teach you about R" by Jenny Bryan and Jim Hester](https://rstats.wtf/debugging-r-code.html) -- it even includes tips for R Markdown debugging;
+
+- [Jenny Bryan's talk "Object of type 'closure' is not subsettable"](https://github.com/jennybc/debugging#readme).
+
+### Beyond R
+
+Sometimes the bug or element to tweak will live outside of R.
+Maybe in some C code you are wrapping, maybe in a CSS file.
+You will therefore have to learn debugging tools for these things too!
+
+## Read tests? Write some for sure
+
+In Patricia Aas' [techniques](https://patricia.no/2018/09/19/reading_other_peoples_code.html) features the idea of writing and running tests to see what's the code is supposed to do.
+She especially mentions integration tests, whereas in R packages you'll mostly find unit tests.
+Those can be useful too to read, especially when they start breaking after your experiments I suppose.
+
+In any case once you have amended a codebase to fix a bug or add a feature, add tests!
+In Kara Woo's talk ["Box plots A case study in debugging and perseverance"](https://www.rstudio.com/resources/rstudioconf-2019/box-plots-a-case-study-in-debugging-and-perseverance/), she explained she added tests.
+In [Jenny Bryan's talk "Object of type 'closure' is not subsettable"](https://github.com/jennybc/debugging#readme) she uses the word "deter" in the part of the talk where she gives such advice.
```

Comment:
this could use a little bit more explanation - I have seen the talk but not recently and I'm left guessing what the reference to "deter" means here.